src/midi_duration.lua

```diff
@@ -0,0 +1,97 @@
+function plugindef()
+    finaleplugin.RequireSelection = true
+    finaleplugin.Author = "Carl Vine"
+    finaleplugin.AuthorURL = "http://carlvine.com"
+    finaleplugin.Copyright = "CC0 https://creativecommons.org/publicdomain/zero/1.0/"
+    finaleplugin.Version = "v1.2"
+    finaleplugin.Date = "2022/06/12"
+    finaleplugin.CategoryTags = "MIDI"
+    finaleplugin.Notes = [[
+    Change the playback START and STOP times for every note in the selected area on one or all layers. 
+    To affect playback "Note Durations" must be enabled under "Playback/Record Options".
+]]
+    return "MIDI Duration", "MIDI Duration", "Change MIDI note start and stop times"
+end
+
+-- RetainLuaState will return global variables:
+-- start_offset, stop_offset and layer_number
+
+function show_error(error_type, param)
+    local errors = {
+        bad_offset = "Offset times must be reasonable,\nsay -9999 to 9999\n(not ",
+        bad_layer_number = "Layer number must be an\ninteger between zero and 4\n(not ",
+    }
+    finenv.UI():AlertNeutral("script: " .. plugindef(), errors[error_type] .. param .. ")")
+end
+
+function user_choices()
+    local current_vert, vert_step = 10, 25
+    local mac_offset = finenv.UI():IsOnMac() and 3 or 0 -- extra y-offset for Mac text box
+    local edit_horiz = 120
+    
+    local dialog = finale.FCCustomWindow()
+    local str = finale.FCString()
+    str.LuaString = plugindef()
+    dialog:SetTitle(str)
+
+    local answer = {}
+    local texts = { -- static text, default value
+        { "Start time:", start_offset or 0 },
+        { "Stop time:", stop_offset or 0 },
+        { "Layer# 1-4 (0 = all):", layer_number or 0 },
+    }
+
+    for i,v in ipairs(texts) do
+        str.LuaString = v[1]
+        local static = dialog:CreateStatic(0, current_vert)
+        static:SetText(str)
+        static:SetWidth(edit_horiz)
+        answer[i] = dialog:CreateEdit(edit_horiz, current_vert - mac_offset)
+        answer[i]:SetInteger(v[2])
+        current_vert = current_vert + vert_step
+    end
+
+    dialog:CreateOkButton()
+    dialog:CreateCancelButton()
+    return dialog:ExecuteModal(nil), 
+        answer[1]:GetInteger(), answer[2]:GetInteger(), answer[3]:GetInteger()
+end
+
+function change_midi_duration()
+    local ok = false
+    ok, start_offset, stop_offset, layer_number = user_choices()
+    if ok ~= finale.EXECMODAL_OK then
+        return
+    end -- user cancelled
+    
+    if start_offset < -9999 or start_offset > 9999 or stop_offset < -9999 or stop_offset > 9999 then
+        show_error("bad_offset", start_offset .. " / " .. stop_offset)
+        return
+    end
+    if layer_number < 0 or layer_number > 4 then
+        show_error("bad_layer_number", layer_number)
+        return
+    end
+    if nil ~= finenv.RetainLuaState then
```

Comment:
Ok, I guess this code review is just going to be me being picky about style.

This line should be:

```lua
finenv.RetainLuaState ~= nil
```

The way you originally wrote it is called a [yoda condition](https://en.wikipedia.org/wiki/Yoda_conditions). While they are valid syntax and there are a few minor advantages of them, they can in general make it harder to read code.